client/src/connection_cache.rs

```diff
@@ -180,101 +182,167 @@ impl ConnectionMap {
-    let mut get_connection_measure = Measure::start("get_connection_measure");
+struct GetConnectionResult {
+    connection: Connection,
+    cache_hit: bool,
+    report_stats: bool,
+    map_timing: u64,
+    lock_timing: u64,
+    connection_cache_stats: Arc<ConnectionCacheStats>,
+    other_stats: Option<(Arc<ClientStats>, ConnectionStats)>,
+}
+
+fn get_or_add_connection(addr: &SocketAddr) -> GetConnectionResult {
     let mut get_connection_map_lock_measure = Measure::start("get_connection_map_lock_measure");
-    let mut map = (*CONNECTION_MAP).lock().unwrap();
+    let map = (*CONNECTION_MAP).read().unwrap();
     get_connection_map_lock_measure.stop();
 
-    if map
+    let mut lock_timing = get_connection_map_lock_measure.as_ms();
+
+    let report_stats = map
         .last_stats
-        .should_update(CONNECTION_STAT_SUBMISSION_INTERVAL)
-    {
-        map.stats.report();
-    }
+        .should_update(CONNECTION_STAT_SUBMISSION_INTERVAL);
 
     let mut get_connection_map_measure = Measure::start("get_connection_hit_measure");
-    let (connection, hit, maybe_stats) = match map.map.get(addr) {
+    let (connection, cache_hit, connection_cache_stats, maybe_stats) = match map.map.get(addr) {
         Some(connection) => {
             let mut stats = None;
             // update connection stats
             if let Connection::Quic(conn) = connection {
                 stats = conn.stats().map(|s| (conn.base_stats(), s));
             }
-            (connection.clone(), true, stats)
+            (connection.clone(), true, map.stats.clone(), stats)
         }
         None => {
             let (_, send_socket) = solana_net_utils::bind_in_range(
                 IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)),
                 VALIDATOR_PORT_RANGE,
             )
             .unwrap();
+
             let connection = if map.use_quic {
                 Connection::Quic(Arc::new(QuicTpuConnection::new(send_socket, *addr)))
             } else {
                 Connection::Udp(Arc::new(UdpTpuConnection::new(send_socket, *addr)))
             };
 
-            map.map.put(*addr, connection.clone());
-            (connection, false, None)
+            // Upgrade to write access by dropping read lock and acquire write lock
+            drop(map);
+            let mut get_connection_map_lock_measure =
+                Measure::start("get_connection_map_lock_measure");
+            let mut map = (*CONNECTION_MAP).write().unwrap();
+            get_connection_map_lock_measure.stop();
+
+            lock_timing = lock_timing.saturating_add(get_connection_map_lock_measure.as_ms());
+
+            // evict a connection if the map is reaching upper bounds
+            while map.map.len() >= MAX_CONNECTIONS {
+                let mut rng = thread_rng();
+                let n = rng.gen_range(0, MAX_CONNECTIONS);
+                if let Some((nth_addr, _)) = map.map.iter().nth(n) {
```

Comment:
Would not this change the algorithm to O(n) once we hit the bound? I think there is a way to still maintain O(1) even when we do random eviction. Another question is, can we simply just evict the latest used address as it is more likely not to be used as the next leader?

Reply:
We typically have 4 consecutive leader slots for a node. Even at that there's no guarantee that the same node will not be picked as the leader for upcoming slots.

This block is there just as a stop gap for time being, just so we don't have unbounded memory growth. I expect this to be updated to some better solution soon.

Reply:
You're right, this would actually make it O(n), unless the tree had the subtree sizes stored...

Reply:
And it does look like the current implementation is O(n).

Reply:
Yes, I checked with @carllin . It seems we actually does not mechanism to disfavor a most recent leader and it can as likely becomes the next leader as others. So the random eviction has lower probability of wrongly evict a leader. To make this algorithm O(1), I think we could use an auxiliary map, index --> address, which can be maintained as elements are added and deleted to the main map.

Reply:
Yes, we need more thoughts on the eviction logic. Anatoly and Trent have some messages on Discord regarding this. We should also take those into account for the algorithm.